OpenRA.Game/GameRules/WeaponInfo.cs

```diff
@@ -177,17 +186,34 @@ public bool IsValidAgainst(FrozenActor victim, Actor firedBy)
 		}
 
 		/// <summary>Applies all the weapon's warheads to the target.</summary>
-		public void Impact(Target target, Actor firedBy, IEnumerable<int> damageModifiers)
+		public void Impact(Target impactTarget, WarheadArgs args)
 		{
+			var world = args.SourceActor.World;
 			foreach (var warhead in Warheads)
 			{
 				var wh = warhead; // force the closure to bind to the current warhead
 
 				if (wh.Delay > 0)
-					firedBy.World.AddFrameEndTask(w => w.Add(new DelayedImpact(wh.Delay, wh, target, firedBy, damageModifiers)));
+			 		world.AddFrameEndTask(w => w.Add(new DelayedImpact(wh.Delay, wh, impactTarget, args)));
 				else
-					wh.DoImpact(target, firedBy, damageModifiers);
+					wh.DoImpact(impactTarget, args);
 			}
 		}
+
+		/// <summary>Applies all the weapon's warheads to the target. Only use for projectile-less, special-case impacts.</summary>
+		public void Impact(Target impactTarget, Actor firedBy, IEnumerable<int> damageModifiers)
```

Comment:
All but two uses of this overload call it as `Impact(Target.FromPos(somePos), someActor, Enumerable.Empty<int>())`. What do you think about changing this to `Impact(WPos pos, Actor firedBy)` and having the other two (`IonCannon`, `TeslaZap`)  construct their own `WarheadArgs` like you've done in the other cases below?